lib/jxl/jpeg/jpeg_data.h

```diff
@@ -217,9 +217,15 @@ struct JPEGData : public Fields {
         has_zero_padding_bit(false) {}
 
   const char* Name() const override { return "JPEGData"; }
+#if JPEGXL_ENABLE_TRANSCODE_JPEG
   // Doesn't serialize everything - skips brotli-encoded data and what is
   // already encoded in the codestream.
   Status VisitFields(Visitor* visitor) override;
+#else
+  Status VisitFields(Visitor* /* visitor */) override {
+    return JXL_ABORT("JPEG transcoding support not enabled");
```

Comment:
either no return, or JXL_FAILURE?

Reply:
If there's no JPEG support built in the idea is that we shouldn't be calling these functions from the API, so I'd prefer an abort here. We are certainly not calling them from the decode.cc.

Reply:
Wouldn't it be better to not have those API functions at all then? If someone would accidentally link an application that does need these functions against a libjxl_dec that doesn't have them, I'd rather have a link time error than a runtime abort, I think...

Reply:
Sure, but why the return ABORT then, instead of just ABORT?

Reply:
clearly I should have used just JXL_ABORT without return..

In no case you will have a runtime abort when using the public API. The JXL_ABORT() here if reached means programming error. If you request the `JXL_DEC_JPEG_RECONSTRUCTION` it will simply never happen, and passing the output buffer will return error.

We can't really remove the functions from the decode.h in the API; unless we make this macro public in the cflags in the .pc file, for systems that use pkg-config, which is not a great solution. Having the functions in the API but not calling them at runtime (or calling them and reacting to the lack of `JXL_DEC_JPEG_RECONSTRUCTION` support) sounds like a valid use-case to me; otherwise you have no way to have a client code that can handle either case (you'd get linker errors).

The question in this patch is whether we want to return error if requesting JXL_DEC_JPEG_RECONSTRUCTION event when not supported or not.

Reply:
Oh, I understand now, thanks for the explanation. I misunderstood and thought it would do a runtime abort when an application tries to do a jpeg reconstruction.

I wouldn't return an error when subscribing to `JXL_DEC_JPEG_RECONSTRUCTION` with a libjxl_dec that doesn't support it. Applications cannot / should not expect all jxl input to be reconstructible to jpeg anyway, so they should not expect that event to certainly happen. By not returning an error, an application will simply consider all inputs to be only decodable to pixels. Returning an error is like more inconvenient for most applications (since they are subscribing to all events at once, and if that fails, they're likely going to give up instead of assuming it's missing jpeg reconstruction support).

That said, it would be good if there is some simple way for applications to know whether this feature is supported or not. In general, as we may add more events to subscribe to later, it would be good for an application to know which events it is actually subscribed to (in the sense that the event can actually happen with the libjxl it is using), instead of only getting a 'SUCCESS' or 'ERROR' after subscribing. Ideally, the SubscribeEvents function would return a bitmask describing the events that were actually subscribed to (which would be a subset of those requested, depending on capabilities of that build/version of the library). Since we can't just change that function now, maybe adding an inspector function that returns the bitmask of currently subscribed-to events would be best?
That way an application can detect that `JXL_DEC_JPEG_RECONSTRUCTION` was not _actually_ subscribed to, if it wants to know that.

Reply:
Yeah, that makes sense. There's the question of forwards/backwards compatibility too. If you run your code against an older version of libjxl (same ABI) some events are not know to that libjxl (you are still able to pass them because for example you compiled your code against a newer header.h, but running against an old one which is legit unless you require symbols not in the other version or when the elf symbol has a different version number). Right now we don't return error for unknown bits in the subscribe-to-events calls, which is why I decided not to return error in this PR. But a new function with the mask of supported events makes sense.